File: solutions/Sliding Window/424_longest-repeating-character-replacement.py
from collections import Counter

class Solution:
    def characterReplacement(self, s: str, k: int) -> int:
        n = len(s)
        freq = {}
        l = 0
        res = 1
        maxf = 0

        for r in range(n):
            freq[s[r]] = freq.get(s[r], 0) + 1
            maxf = max(maxf, freq[s[r]])

            if (r - l + 1) - maxf > k:
                freq[s[l]] -= 1
                l += 1

            res = max(res, r - l + 1)

        return res

# Time Complexity: O(n)
# Space Complexity: O(1)


class Solution:
    def characterReplacement(self, s: str, k: int) -> int:
        n = len(s)
        freq = Counter()
        l = 0
        res = 1

        for r in range(n):
            freq[s[r]] += 1

            while (r - l + 1) - max(freq.values()) > k:
                freq[s[l]] -= 1
                l += 1

            res = max(res, r - l + 1)

        return res

# Time Complexity: O(n)
# Space Complexity: O(1)

File: C++/Sliding Window/424_longest-repeating-character-replacement.cpp
#include <unordered_map>
#include <string>

using namespace std;

class Solution {
public:
    int characterReplacement(string s, int k) {
        size_t n = s.size();
        unordered_map<char, int> freq;

        int res = 0;
        int l = 0;
        for (int r = 0; r < n; r++) {
            freq[s[r]]++;

            while ((r - l + 1) - getMax(freq) > k) freq[s[l++]]--;

            res = max(res, r - l + 1);
        }

        return res;
    }

    int getMax(const unordered_map<char, int>& freq) {
        int res = 0;
        for (auto [ch, f]: freq) res = max(res, f);

        return res;
    }
};


class Solution {
public:
    int characterReplacement(string s, int k) {
        size_t n = s.size();
        int maxf = 0;
        unordered_map<char, int> freq;

        int res = 0;
        int l = 0;
        for (int r = 0; r < n; r++) {
            freq[s[r]]++;
            maxf = max(maxf, freq[s[r]]);

            if ((r - l + 1) - maxf > k) freq[s[l++]]--;

            res = max(res, r - l + 1);
        }

        return res;
    }
};